### PR TITLE
Backport GPR#1550 on 4.06

### PR DESCRIPTION
**Pull request by trefis**
As I said, the GPR came after the big rewrite of parmatch, so it didn't apply cleanly.
I tried to minimize the amount of extra code added on top of #1550 , but I had to pull in a modified version of `simplify_first_col`. Making sure it's correct should be fairly easy, but I encourage people to double check.

Also, the ambiguous variable check looked very different and here too I encourage you to make sure the backport is correct.

### CODE REVIEW

File: typing/parmatch.ml [hunk 1]
```diff
@@ -44,6 +44,151 @@ let omega_list l = List.map (fun _ -> omega) l
...
+let simplify_head_pat p k =
+  let rec simplify_head_pat p k =
+    match p.pat_desc with
+    | Tpat_alias (p,_,_) -> simplify_head_pat p k
+    | Tpat_var (_,_) -> omega :: k
+    | Tpat_or (p1,p2,_) -> simplify_head_pat p1 (simplify_head_pat p2 k)
+    | _ -> p :: k
+  in simplify_head_pat p k
+
+let rec simplified_first_col = function
```
Comment (gasche) [inline typing/parmatch.ml:128]: why did you change the name of this function compared to trunk, which uses `simplify` instead of `simplified`? I have no strong opinion on which is best, but note that you use `simplify` for patterns above and I think that using different function names in trunk and 4.06 is not a good idea.

File: typing/parmatch.ml [hunk 2]
```diff
@@ -44,6 +44,151 @@ let omega_list l = List.map (fun _ -> omega) l
...
+          | Const_nativeint _
+          | Const_float _
+          | Const_string _), _ -> false
+      end
+    | Tpat_tuple l1, Tpat_tuple l2 -> List.length l1 = List.length l2
+    | Tpat_record ((_, lbl1, _) :: _, _), Tpat_record ((_, lbl2, _) :: _, _) ->
+      Array.length lbl1.lbl_all = Array.length lbl2.lbl_all
+    | Tpat_any, _
+    | _, Tpat_any
+    | Tpat_record ([], _), Tpat_record ([], _)
```
Comment (gasche) [inline typing/parmatch.ml:169]: This line is still wrong -- see the recent comments in #1550. If you have an opinion about it, feel free to comment, but otherwise please fix it, or remove it, or turn this `[]` case into an `assert false`.

File: typing/parmatch.ml [hunk 3]
```diff
@@ -1049,26 +1208,36 @@ let rec satisfiable pss qs = match pss with
...
-              List.exists
-                (fun (p,pss) ->
-                  not (is_absent_pat p) &&
-                  satisfiable pss (simple_match_args p omega @ qs))
-                constrs
-            else
-              satisfiable (filter_extra pss) qs
+        if not (all_coherent (simplified_first_col pss)) then
+          false
+        else begin
```
Comment (gasche) [inline typing/parmatch.ml:1213]: Thanks for using `begin`..`end` :-)

**Review by gasche — COMMENTED** [review]
I just did a review -- I propose some changes, but overall the patch looks nice and correct to me. Due to the absence of @trefis refactoring in this branch, we do a bit of redundant computation (simplification is not shared between coherence testing and the computation in each case), but I think that is fine.

**Comment by gasche** [comment]
Also, I think we need some sort of Changes entry -- it should be easy for others to audit what went into the bugfix release. Could you at least add this PR number and the MPR regression number to the MPR#7661, GPR#1459 Changes entry?

File: typing/parmatch.ml (hunk 2 above)
Comment (trefis) [reply typing/parmatch.ml:169]: Allow me to start things off by quoting you:
> This is of no consequence for the compiler given that empty record patterns are rejected by the type-checker anyway, so the PR is safe

I do not mind changing this *in trunk*, allowing empty record patterns at the same time, and giving them the semantic you have in mind (which I'll agree is nicer than the one I initially considered).
But I'm not keen of making this backport diverge even more from the original GPR.

I'll do the change if you insist though.

File: typing/parmatch.ml (hunk 1 above)
Comment (trefis) [reply typing/parmatch.ml:128]: I chose to use a different name because the function is different:
- in trunk we return the matrix, with its first column simplified
- here we only return the simplified first column

**Comment by trefis** [comment]
It does indeed need a Changes entry, I just wasn't sure if we wanted to move the initial entry to the 4.06 section, or duplicate it, or ...

What do you think would be best?

**Comment by gasche** [comment]
I would add the following Changes entry to the 4.06 section:

```
- MPR#NNNN, GPR#1587: Make pattern matching warnings more robust
  to ill-typed columns; this is a backport of GPR#1550 from 4.07+dev
  (Thomas Refis, review by Gabriel Scherer, report by <mantis reporter>)
```

File: typing/parmatch.ml (hunk 2 above)
Comment (gasche) [reply typing/parmatch.ml:169]: I don't think that empty record patterns should be allowed without careful consideration, because doing type-inference for them is delicate (just as for empty sums) unless there is a propagated type annotation.

I don't like the idea of having wrong dead code in the compiler. We know it's not going to break today, but it may break later, confuse people, or have various unplanned costs. So I'm not excited at the idea of committing more wrong code. In this specific case I think that the best course of action is to fix it someway (someway safe!) in this PR, and then also send the fix against trunk to avoid divergence.

File: typing/parmatch.ml (hunk 1 above)
Comment (gasche) [reply typing/parmatch.ml:128]: Ah, I see. That's reasonable.

**Comment by gasche** [comment]
The mantis issue is [MPR#7713](https://caml.inria.fr/mantis/view.php?id=7713) and the reporter is Andreas Hauptmann.

**Comment by gasche** [comment]
I would be happy if we could get this merged quickly to consider a new round on testing on the 4.06 branch. @trefis, would it help if I did some changes myself (the empty record silliness and a Changes entry)? I don't mind working on this later today.

**Comment by trefis** [comment]
I won’t have access to a computer before Monday, so if you want it merge before then feel free to do the changes you want.

For the record, I still disagree about the empty record case: I feel like this should be changed in trunk, not here.
But I understand that those who are away are always wrong, so do what you will.

**Comment by gasche** [comment]
I performed the changes in [my personal backport-1550 branch](https://github.com/ocaml/ocaml/compare/4.06...gasche:backport-1550) but it looks like I cannot force-push to your clone (fair enough; it works for some fork-specific branches, but I don't really understand why), so I may create a new PR for this.

**Comment by gasche** [comment]
Merged as #1595.
